### ISSUE TITLE
prow: trigger: don't start postsubmits on branch deletion

### ISSUE THREAD

**Issue by stevekuznetsov** [kubernetes/test-infra#9798]
When a branch was deleted, we saw postsubmits start for it. Logs from `hook`:

```
{"component":"hook","event-GUID":"8ad7bb2a-ce1d-11e8-8900-358d64bd12bf","event-type":"push","head":"0000000000000000000000000000000000000000","level":"info","msg":"Push event.","org":"openshift","ref":"refs/heads/master-dockerless-build-temporary","repo":"origin","time":"2018-10-12T12:51:31Z"}
{"component":"hook","event-GUID":"8ad7bb2a-ce1d-11e8-8900-358d64bd12bf","event-type":"push","head":"0000000000000000000000000000000000000000","job":"ami_build_origin_int_rhel_build","level":"info","msg":"Creating a new prowjob.","name":"8b05087e-ce1d-11e8-aac2-0a58ac102050","org":"openshift","plugin":"trigger","ref":"refs/heads/master-dockerless-build-temporary","repo":"origin","time":"2018-10-12T12:51:31Z","type":"postsubmit"}
{"client":"kube","component":"hook","level":"debug","msg":"CreateProwJob({\"kind\":\"ProwJob\",\"apiVersion\":\"prow.k8s.io/v1\",\"metadata\":{\"name\":\"8b05087e-ce1d-11e8-aac2-0a58ac102050\",\"creationTimestamp\":null,\"labels\":{\"event-GUID\":\"8ad7bb2a-ce1d-11e8-8900-358d64bd12bf\",\"master\":\"ci.openshift.redhat.com\",\"prow.k8s.io/job\":\"ami_build_origin_int_rhel_build\",\"prow.k8s.io/refs.org\":\"openshift\",\"prow.k8s.io/refs.repo\":\"origin\",\"prow.k8s.io/type\":\"postsubmit\"},\"annotations\":{\"prow.k8s.io/job\":\"ami_build_origin_int_rhel_build\"}},\"spec\":{\"type\":\"postsubmit\",\"agent\":\"jenkins\",\"job\":\"ami_build_origin_int_rhel_build\",\"refs\":{\"org\":\"openshift\",\"repo\":\"origin\",\"base_ref\":\"master-dockerless-build-temporary\",\"base_sha\":\"0000000000000000000000000000000000000000\"}},\"status\":{\"startTime\":\"2018-10-12T12:51:31Z\",\"state\":\"triggered\",\"prev_report_state\":\"\"}})","time":"2018-10-12T12:51:31Z"}
{"component":"hook","event-GUID":"8ad7bb2a-ce1d-11e8-8900-358d64bd12bf","event-type":"push","head":"0000000000000000000000000000000000000000","job":"branch-ci-openshift-origin-master-images","level":"info","msg":"Creating a new prowjob.","name":"8b06f7da-ce1d-11e8-aac2-0a58ac102050","org":"openshift","plugin":"trigger","ref":"refs/heads/master-dockerless-build-temporary","repo":"origin","time":"2018-10-12T12:51:31Z","type":"postsubmit"}
{"client":"kube","component":"hook","level":"debug","msg":"CreateProwJob({\"kind\":\"ProwJob\",\"apiVersion\":\"prow.k8s.io/v1\",\"metadata\":{\"name\":\"8b06f7da-ce1d-11e8-aac2-0a58ac102050\",\"creationTimestamp\":null,\"labels\":{\"artifacts\":\"images\",\"event-GUID\":\"8ad7bb2a-ce1d-11e8-8900-358d64bd12bf\",\"prow.k8s.io/job\":\"branch-ci-openshift-origin-master-images\",\"prow.k8s.io/refs.org\":\"openshift\",\"prow.k8s.io/refs.repo\":\"origin\",\"prow.k8s.io/type\":\"postsubmit\"},\"annotations\":{\"prow.k8s.io/job\":\"branch-ci-openshift-origin-master-images\"}},\"spec\":{\"type\":\"postsubmit\",\"agent\":\"kubernetes\",\"cluster\":\"default\",\"job\":\"branch-ci-openshift-origin-master-images\",\"refs\":{\"org\":\"openshift\",\"repo\":\"origin\",\"base_ref\":\"master-dockerless-build-temporary\",\"base_sha\":\"0000000000000000000000000000000000000000\"},\"pod_spec\":{\"containers\":[{\"name\":\"\",\"image\":\"ci-operator:latest\",\"command\":[\"ci-operator\"],\"args\":[\"--artifact-dir=$(ARTIFACTS)\",\"--give-pr-author-access-to-namespace=true\",\"--promote\",\"--target=[images]\",\"--target=artifacts\",\"--target=base\"],\"env\":[{\"name\":\"CONFIG_SPEC\",\"valueFrom\":{\"configMapKeyRef\":{\"name\":\"ci-operator-configs\",\"key\":\"openshift-origin-master.yaml\"}}}],\"resources\":{\"limits\":{\"cpu\":\"500m\"},\"requests\":{\"cpu\":\"10m\"}},\"imagePullPolicy\":\"Always\"}],\"serviceAccountName\":\"ci-operator\"},\"decoration_config\":{\"timeout\":14400000000000,\"grace_period\":15000000000,\"utility_images\":{\"clonerefs\":\"clonerefs:latest\",\"initupload\":\"initupload:latest\",\"entrypoint\":\"entrypoint:latest\",\"sidecar\":\"sidecar:latest\"},\"gcs_configuration\":{\"bucket\":\"origin-ci-test\",\"path_strategy\":\"single\",\"default_org\":\"openshift\",\"default_repo\":\"origin\"},\"gcs_credentials_secret\":\"gcs-publisher-credentials\",\"skip_cloning\":true}},\"status\":{\"startTime\":\"2018-10-12T12:51:31Z\",\"state\":\"triggered\",\"prev_report_state\":\"\"}})","time":"2018-10-12T12:51:31Z"}
{"component":"hook","event-GUID":"8ad7bb2a-ce1d-11e8-8900-358d64bd12bf","event-type":"push","head":"0000000000000000000000000000000000000000","job":"test_branch_origin_extended_conformance_azure","level":"info","msg":"Creating a new prowjob.","name":"8b07c4b0-ce1d-11e8-aac2-0a58ac102050","org":"openshift","plugin":"trigger","ref":"refs/heads/master-dockerless-build-temporary","repo":"origin","time":"2018-10-12T12:51:31Z","type":"postsubmit"}
{"client":"kube","component":"hook","level":"debug","msg":"CreateProwJob({\"kind\":\"ProwJob\",\"apiVersion\":\"prow.k8s.io/v1\",\"metadata\":{\"name\":\"8b07c4b0-ce1d-11e8-aac2-0a58ac102050\",\"creationTimestamp\":null,\"labels\":{\"event-GUID\":\"8ad7bb2a-ce1d-11e8-8900-358d64bd12bf\",\"master\":\"ci.openshift.redhat.com\",\"prow.k8s.io/job\":\"test_branch_origin_extended_conformance_azure\",\"prow.k8s.io/refs.org\":\"openshift\",\"prow.k8s.io/refs.repo\":\"origin\",\"prow.k8s.io/type\":\"postsubmit\"},\"annotations\":{\"prow.k8s.io/job\":\"test_branch_origin_extended_conformance_azure\"}},\"spec\":{\"type\":\"postsubmit\",\"agent\":\"jenkins\",\"job\":\"test_branch_origin_extended_conformance_azure\",\"refs\":{\"org\":\"openshift\",\"repo\":\"origin\",\"base_ref\":\"master-dockerless-build-temporary\",\"base_sha\":\"0000000000000000000000000000000000000000\"}},\"status\":{\"startTime\":\"2018-10-12T12:51:31Z\",\"state\":\"triggered\",\"prev_report_state\":\"\"}})","time":"2018-10-12T12:51:31Z"}
```

This is weird. When I deleted a branch manually, I see the appropriate header:

```
X-GitHub-Event: delete
```

Prow grabs the header here:

https://github.com/kubernetes/test-infra/blob/3b0b2395a41335d0a29e4071837267651b99977a/prow/github/webhooks.go#L44-L48

And we only issue a push event here:

https://github.com/kubernetes/test-infra/blob/3b0b2395a41335d0a29e4071837267651b99977a/prow/hook/server.go#L133-L141

So it's not clear to me how this is happening, but this is a correctness bug in Prow.

/area prow
/area prow/hook
/kind bug
/priority critical-urgent
/assign

**Comment by stevekuznetsov** [comment]
This leads to a ProwJob with weird refs:

```json
"refs": {
  "base_ref": "master-dockerless-build-temporary",
  "base_sha": "0000000000000000000000000000000000000000",
  "org": "openshift",
  "repo": "origin"
}
```

<details><summary>full ProwJob JSON</summary>

```json
{
  "apiVersion": "prow.k8s.io/v1",
  "kind": "ProwJob",
  "metadata": {
    "annotations": {
      "prow.k8s.io/job": "branch-ci-openshift-origin-master-images"
    },
    "creationTimestamp": null,
    "labels": {
      "artifacts": "images",
      "event-GUID": "8ad7bb2a-ce1d-11e8-8900-358d64bd12bf",
      "prow.k8s.io/job": "branch-ci-openshift-origin-master-images",
      "prow.k8s.io/refs.org": "openshift",
      "prow.k8s.io/refs.repo": "origin",
      "prow.k8s.io/type": "postsubmit"
    },
    "name": "8b06f7da-ce1d-11e8-aac2-0a58ac102050"
  },
  "spec": {
    "agent": "kubernetes",
    "cluster": "default",
    "decoration_config": {
      "gcs_configuration": {
        "bucket": "origin-ci-test",
        "default_org": "openshift",
        "default_repo": "origin",
        "path_strategy": "single"
      },
      "gcs_credentials_secret": "gcs-publisher-credentials",
      "grace_period": 15000000000,
      "timeout": 14400000000000,
      "utility_images": {
        "clonerefs": "clonerefs:latest",
        "entrypoint": "entrypoint:latest",
        "initupload": "initupload:latest",
        "sidecar": "sidecar:latest"
      }
    },
    "job": "branch-ci-openshift-origin-master-images",
    "pod_spec": {
      "containers": [
        {
          "args": [
            "--artifact-dir=$(ARTIFACTS)",
            "--give-pr-author-access-to-namespace=true",
            "--promote",
            "--target=[images]",
            "--target=artifacts",
            "--target=base"
          ],
          "command": [
            "ci-operator"
          ],
          "env": [
            {
              "name": "CONFIG_SPEC",
              "valueFrom": {
                "configMapKeyRef": {
                  "key": "openshift-origin-master.yaml",
                  "name": "ci-operator-configs"
                }
              }
            }
          ],
          "image": "ci-operator:latest",
          "imagePullPolicy": "Always",
          "name": "",
          "resources": {
            "limits": {
              "cpu": "500m"
            },
            "requests": {
              "cpu": "10m"
            }
          }
        }
      ],
      "serviceAccountName": "ci-operator"
    },
    "refs": {
      "base_ref": "master-dockerless-build-temporary",
      "base_sha": "0000000000000000000000000000000000000000",
      "org": "openshift",
      "repo": "origin"
    },
    "type": "postsubmit"
  },
  "status": {
    "prev_report_state": "",
    "startTime": "2018-10-12T12:51:31Z",
    "state": "triggered"
  }
}
```

</details>

**Comment by stevekuznetsov** [comment]
Oh, before we get the `delete` event, we get:

```
X-GitHub-Event: push
...
{
  "ref": "refs/heads/skuznets/test",
  "before": "1cac8bc5f57b116911e8aedbabf6f0456dd51de6",
  "after": "0000000000000000000000000000000000000000",
  "created": false,
  "deleted": true,
  "forced": false,
  "base_ref": null,
  "compare": "https://github.com/openshift/ci-operator/compare/1cac8bc5f57b...000000000000",
  "commits": [

  ],
  "head_commit": null,
...
```

WTF GitHub. I imagine this is the same issue with the Slack integration on branch deletes.

**Comment by cjwagner** [comment]
Can we use the `"deleted": true,` field to reliably filter out these events?

**Comment by stevekuznetsov** [comment]
@cjwagner 

Yes, that was my approach in the PR.

**Comment by alvaroaleman** [comment]
Fixed in #10293 

/close
